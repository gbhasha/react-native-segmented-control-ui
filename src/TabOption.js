import React from 'react'
import { View, ViewPropTypes,  TouchableOpacity, StyleSheet,  Text } from 'react-native'
import PropTypes from 'prop-types'

const TabOption = ({
    isTabActive, index, badge, text,
    firstTabStyle, lastTabStyle,
    tabStyle, activeTabStyle,
    tabTextStyle, activeTabTextStyle,
    tabBadgeContainerStyle, activeTabBadgeContainerStyle,
    tabBadgeStyle, activeTabBadgeStyle,
    onTabPress, textNumberOfLines,
    allowFontScaling,
    accessible,
    accessibilityLabel,
    testID,
}) => {
    return (
        <TouchableOpacity style={[
            styles.tabStyle,
            tabStyle,
            isTabActive ? [styles.activeTabStyle, activeTabStyle] : {},
            firstTabStyle,
            lastTabStyle]}
            testID={testID}
            accessible={accessible}
            accessibilityLabel={accessibilityLabel}
            accessibilityTraits={isTabActive ? "selected" : "button"}
            accessibilityComponentType={"button"}
            onPress={() => onTabPress(index)}
            activeOpacity={1}>
            <View style={{ flexDirection: "row" }}>
                <Text style={[
                    styles.tabTextStyle,
                    tabTextStyle,
                    isTabActive ? [styles.activeTabTextStyle, activeTabTextStyle] : {}]}
                    numberOfLines={textNumberOfLines}
                    allowFontScaling={allowFontScaling}
                    ellipsizeMode="tail">
                    {text}
                </Text>
                {
                    badge !== undefined &&
                        <View style={[
                            styles.tabBadgeContainerStyle,
                            tabBadgeContainerStyle,
                            isTabActive ? [styles.activeTabBadgeContainerStyle, activeTabBadgeContainerStyle] : {}]}>
                            <Text style={[
                                styles.tabBadgeStyle,
                                tabBadgeStyle,
                                isTabActive ? [styles.activeTabBadgeStyle, activeTabBadgeStyle] : {}]}
                                allowFontScaling={allowFontScaling}>
                                {badge}
                            </Text>
                        </View>
                }
            </View>
        </TouchableOpacity>
    )
}

const styles = StyleSheet.create({
    tabStyle: {
        paddingVertical: 5,
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        borderLeftColor: '#007AFF',
        borderLeftWidth: 1,
        backgroundColor: 'white',
    },
    activeTabStyle: {
        backgroundColor: '#007AFF'
    },
    tabTextStyle: {
        color: '#007AFF'
    },
    activeTabTextStyle: {
        color: 'white'
    },
    tabBadgeContainerStyle: {
        borderRadius: 20,
        backgroundColor: 'red',
        paddingLeft: 5,
        paddingRight: 5,
        marginLeft: 5,
        marginBottom: 3
    },
    activeTabBadgeContainerStyle: {
        backgroundColor: 'white'
    },
    tabBadgeStyle: {
        color: 'white',
        fontSize: 11,
        fontWeight: 'bold'
    },
    activeTabBadgeStyle: {
        color: 'black'
    }
})

TabOption.propTypes = {
    accessibilityLabel: PropTypes.string,
    accessible: PropTypes.bool,
    activeTabBadgeContainerStyle: Text.propTypes.style,
    activeTabBadgeStyle: Text.propTypes.style,
    activeTabStyle: ViewPropTypes.style,
    activeTabTextStyle: Text.propTypes.style,
    allowFontScaling: PropTypes.bool,
    badge: PropTypes.number,
    firstTabStyle: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
    index: PropTypes.number,
    isTabActive: PropTypes.bool,
    lastTabStyle: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
    onTabPress: PropTypes.func,
    tabBadgeContainerStyle: Text.propTypes.style,
    tabBadgeStyle: Text.propTypes.style,
    tabStyle: ViewPropTypes.style,
    tabTextStyle: Text.propTypes.style,
    tabsContainerStyle: ViewPropTypes.style,
    testID: PropTypes.string,
    text: PropTypes.string,
    textNumberOfLines: PropTypes.number,
  }

export default TabOption