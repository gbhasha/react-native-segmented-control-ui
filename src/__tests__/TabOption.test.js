// __tests__/Intro-test.js
import React from 'react';
import TabOption from '../TabOption';

import renderer from 'react-test-renderer';

describe('TabOption', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  });

  it('renders correctly', () => {
    const tree = renderer.create(<TabOption />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should apply active tab styles when isTabActive is set to true', () => {
    const activeTabStyle = {backgroundColor: '#0000FF'}
    const tree = renderer.create(<TabOption
                    activeTabStyle={activeTabStyle}
                    isTabActive />).toJSON();
    expect(tree.props.style.backgroundColor).toBe(activeTabStyle.backgroundColor)
  });

  it('should call the onPress handler when tapped', () => {
    const tabIndex = 1
    const mockFn = jest.fn()
    const tree = renderer.create(<TabOption
                    testID='tab'
                    index={tabIndex}
                    onTabPress={mockFn} />).toJSON()
    jest.mock('TouchableOpacity', () => {
      const jestReactNative = require('jest-react-native')
      return jestReactNative('TouchableOpacity')
    })
    tree.props.onPress();
    expect(mockFn).toHaveBeenCalledWith(tabIndex)

  });

  it('should render the badge along with tab text when badge is passed in props ', () => {
    const tree = renderer.create(<TabOption badge={5}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should apply the passed badgeStyle when tab is active ', () => {
    const styles = {
      activeTabStyle : {backgroundColor: '#0000FF'},
      activeTabBadgeStyle : {color: '#CCCCCC'},
      activeTabBadgeContainerStyle : {backgroundColor: '#00FFFF'}
    }

    const tree = renderer.create(<TabOption badge={5}
                    activeTabStyle={styles.activeTabStyle}
                    activeTabBadgeContainerStyle={styles.activeTabBadgeContainerStyle}
                    activeTabBadgeStyle={styles.activeTabBadgeStyle}
                    isTabActive />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})


