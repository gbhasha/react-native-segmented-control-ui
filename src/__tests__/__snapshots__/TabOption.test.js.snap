// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabOption renders correctly 1`] = `
<View
  accessibilityComponentType="button"
  accessibilityTraits="button"
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderLeftColor": "#007AFF",
      "borderLeftWidth": 1,
      "flex": 1,
      "justifyContent": "center",
      "opacity": 1,
      "paddingVertical": 5,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#007AFF",
          },
          undefined,
          Object {},
        ]
      }
    />
  </View>
</View>
`;

exports[`TabOption should apply the passed badgeStyle when tab is active  1`] = `
<TouchableOpacity
  accessibilityComponentType="button"
  accessibilityTraits="selected"
  activeOpacity={1}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderLeftColor": "#007AFF",
        "borderLeftWidth": 1,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 5,
      },
      undefined,
      Array [
        Object {
          "backgroundColor": "#007AFF",
        },
        Object {
          "backgroundColor": "#0000FF",
        },
      ],
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#007AFF",
          },
          undefined,
          Array [
            Object {
              "color": "white",
            },
            undefined,
          ],
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "red",
            "borderRadius": 20,
            "marginBottom": 3,
            "marginLeft": 5,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
          undefined,
          Array [
            Object {
              "backgroundColor": "white",
            },
            Object {
              "backgroundColor": "#00FFFF",
            },
          ],
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 11,
              "fontWeight": "bold",
            },
            undefined,
            Array [
              Object {
                "color": "black",
              },
              Object {
                "color": "#CCCCCC",
              },
            ],
          ]
        }
      >
        5
      </Text>
    </View>
  </View>
</TouchableOpacity>
`;

exports[`TabOption should render the badge along with tab text when badge is passed in props  1`] = `
<TouchableOpacity
  accessibilityComponentType="button"
  accessibilityTraits="button"
  activeOpacity={1}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderLeftColor": "#007AFF",
        "borderLeftWidth": 1,
        "flex": 1,
        "justifyContent": "center",
        "paddingVertical": 5,
      },
      undefined,
      Object {},
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#007AFF",
          },
          undefined,
          Object {},
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "red",
            "borderRadius": 20,
            "marginBottom": 3,
            "marginLeft": 5,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
          undefined,
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 11,
              "fontWeight": "bold",
            },
            undefined,
            Object {},
          ]
        }
      >
        5
      </Text>
    </View>
  </View>
</TouchableOpacity>
`;
